arbor/network.cpp

```diff
@@ -0,0 +1,1551 @@
+#include <arbor/common_types.hpp>
+#include <arbor/network.hpp>
+
+#include <Random123/threefry.h>
+#include <Random123/boxmuller.hpp>
+#include <Random123/uniform.hpp>
+
+#include <algorithm>
+#include <cmath>
+#include <cstdint>
+#include <memory>
+#include <optional>
+#include <type_traits>
+#include <variant>
+#include <vector>
+
+#include "network_impl.hpp"
+
+namespace arb {
+
+namespace {
+
+// Partial seed to use for network_value and network_selection generation.
+// Different seed for each type to avoid unintentional correlation.
+enum class network_seed : unsigned {
+    selection_random = 2058443,
+    value_uniform = 48202,
+    value_normal = 8405,
+    value_truncated_normal = 380237,
+};
+
+
+double uniform_rand_from_key_pair(std::array<unsigned, 2> seed,
+    network_hash_type key_a,
+    network_hash_type key_b) {
+    using rand_type = r123::Threefry2x64;
+    const rand_type::ctr_type seed_input = {{seed[0], seed[1]}};
+
+    const rand_type::key_type key = {{std::min(key_a, key_b), std::max(key_a, key_b)}};
+    rand_type gen;
+    return r123::u01<double>(gen(seed_input, key)[0]);
+}
+
+double normal_rand_from_key_pair(std::array<unsigned, 2> seed,
+    std::uint64_t key_a,
+    std::uint64_t key_b) {
+    using rand_type = r123::Threefry2x64;
+    const rand_type::ctr_type seed_input = {{seed[0], seed[1]}};
+
+    const rand_type::key_type key = {{std::min(key_a, key_b), std::max(key_a, key_b)}};
+    rand_type gen;
+    const auto rand_num = gen(seed_input, key);
+    return r123::boxmuller(rand_num[0], rand_num[1]).x;
+}
+
+struct network_selection_all_impl: public network_selection_impl {
+    bool select_connection(const network_full_site_info& src,
+        const network_full_site_info& dest) const override {
+        return true;
+    }
+
+    bool select_source(cell_kind kind,
+        cell_gid_type gid,
+        const std::string_view& label) const override {
+        return true;
+    }
+
+    bool select_destination(cell_kind kind,
+        cell_gid_type gid,
+        const std::string_view& label) const override {
```

Comment:
I wasn't actually aware, that it's a const-view only. So yeah, it's  it might contra-intuitive, but then again it might prevent misunderstanding for others, who don't know either 😉 . But I'll change it if you prefer, I don't mind either way.

Reply:
As much as I want to love `std::string_view` it seems always a bit awkward.

Reply:
The main issue I had here is that `string_view` is a reference-like thing which gets passed by ref.